scripts/check_whitespace.py

```diff
@@ -0,0 +1,124 @@
+import sys
+
+# Bad whitespace is
+# - a TAB (anywhere),
+# - a SPACE before CR,
+# - a SPACE before NL, or
+# - a SPACE as last character.
+BAD_WHITESPACE = re.compile(r'\t| \r| \n| $')
+
+
+def has_bad_whitespace(path):
+    """
+    Returns whether the file at |path| has bad whitespace.
+    """
+    if os.path.isfile(path):
+        with open(path, 'r') as handle:
+            for line in handle:
+                m = BAD_WHITESPACE.search(line)
+                if m:
+                    return True
+
+    return False
+
+
+def main():
+    # Process the command line.
+    short_opts = "he:"
+    long_opts = []
+    try:
+        opts, args = getopt.getopt(sys.argv[1:], short_opts, long_opts)
+    except getopt.GetoptError as ex:
+        print(f"ERROR: {ex}")
+        sys.exit(1)
+
+    # Process the found options.
+    excludes = []
+    for opt, optval in opts:
+        if opt == '-h':
+            print("Usage: check_whitespace [options] [ROOT-DIR ...]")
+            print()
+            print("Checks source files for bad whitespace. Recursively examines all")
+            print("supplied ROOT-DIR, or the current directory if no root is provided.")
+            print()
+            print("Options:")
+            print("  -h       Print this help text")
+            print("  -e TEXT  Exclude source files with TEXT in their path")
+            print("           This option may be used multiple times")
+            print()
+            print("Have a nice day!")
+            sys.exit(0)
+
+        if opt == '-e':
+            excludes.append(optval)
+            continue
+
+        # Shouldn't happen due to getopt checking.
+        assert False, f"Unexpected option ({(opt, optval)} found."
+
+    # Setup root directories.
+    cur_dir = os.getcwd()
+    if args:
+        tops = [os.path.join(cur_dir, arg) for arg in args]
+    else:
+        tops = [cur_dir]
+
+    # Process files.
+    count = 0
+    offending_files = []
+    found_errors = False
+    for top in tops:
+        for root, dirs, files in os.walk(top):
+            for f in files:
+                # Skip non-source files.
+                if not f.lower().endswith(('.py', '.lua', '.h', '.cpp', '.cc', '.c')):
+                    continue
+
+                # Skip files with excluded patterns.
+                path = os.path.join(root, f)
+                if any(excl in path for excl in excludes):
+                    continue
+
+                # Check the file.
+                count += 1
+                try:
+                    if has_bad_whitespace(path):
+                        found_errors = True
+                        offending_files.append(path)
+
+                except UnicodeDecodeError:
+                    print(f"ERROR: File {path} has Unicode errors.")
+                    found_errors = True
+
+    # Report files with bad whitespace.
+    print('Checked {} files'.format(count))
+    if offending_files:
+        print('Found files with bad whitespace:')
```

Comment:
Is there a chance it could also print the line(s) it was found on/what the whitespace issue was?

Reply:
It's feasible of course.

I did consider it a bit, but didn't know what to do if you have a lot of these problems. I feel there isn't much point in listing all, for eg the previous version of the two changed language files.

The kind of problem could probably be useful.

Reply:
Yeah, Luacheck's copy does output the offending issue and the line:column as far as I remember. At the very least I agree the kind of issue will be very useful to especially someone new to our code.

Reply:
The file is enough to find it, I wouldn't print the lines.